### DISCUSSION TITLE
authStore.isValid is randomly false

### DISCUSSION THREAD

**Discussion by kkosmowski** [pocketbase/pocketbase#4758] [Q&A] [answered]
I have added PocketBase to my Next.js app, I have to log in again after making changes to app during development.

I've noticed I very often get logged out right after I logged in. 

I have no issue sharing my app:
https://github.com/kkosmowski/next-social-app

Here's sum up of how my auth works:

1. client side makes request
sending POST with credentials to my server-side route handler

2. server side authorises user with password, sets cookies and responds back
```ts
    const authData = await pb.users.authWithPassword(payload.email, payload.password);
    pb.authStore.save(authData.token, authData.record);
    const data = mapUserRecordToUser(authData.record);

    cookies().set(TOKEN_COOKIE_KEY, authData.token);
    const response = NextResponse.json(data);
    response.cookies.set(TOKEN_COOKIE_KEY, authData.token);
    return response;
```

* I have read that I don't need to do `pb.authStore.save()`, it in fact does not change anything. It's an old line of code.

3. client side handles the response and moves user to main page.
Client side uses AuthProvider to keep the state of session. 

**There is also an analogical class instance for storing session on server side, but it will be empty until invoking `authRefresh()` for the first time.** This is why in the video below it will log `isLoggedIn: false`.

4. immediately after signing in OR even after page refresh, when checking `pb.authStore.isValid`, at times it will be `false` for no reason.

In my case it is one of three:
- trying to "like" a post
- trying to navigate to "me page"
- trying to log out

Each of these does the same:
1. route handler checks if there is user data cached (**initially it is not**)
2. if yes – cache the data and return it, otherwise check `pb.authStore.isValid`
3. if valid, call `authRefresh()` and save the data for future checks
4. else log the user out

```ts
    if (this.data.isLoggedIn && this.isDataFresh()) {
      return;
    }

    if (pb.authStore.isValid) {
      const authData = await pb.users.authRefresh();

      this.data = {
        user: mapUserRecordToUser(authData.record),
        token: authData.token,
        isLoggedIn: true,
      };
      this.updatedAt = new Date();
    } else {
      this.logout();
    }
```

Here's a short video presentation:

https://github.com/pocketbase/pocketbase/assets/63412876/926d2274-5ad7-465f-b569-377137aa1b53

I have added few console logs for this video – after signing in, to show that it was successful, and right before getting falsy `pb.authStore.isValid`. These console logs are available on `pocketbase-debug` branch.

Perhaps I am doing something wrong or missing a crucial step. Any help will be appreciated.
Thanks 🙌

**Comment by ganigeorgiev** [comment, marked as answer]
There is too much going on in the linked repo to understand what actually is happening but I don't think this is an issue with the JS SDK and most likely there is something wrong with your Next.js integration.

`pb.authStore` represents a local auth state. `pb.authStore.isValid` only loosely checks whether there is a token loaded into the store and it is not expired (by extracting the `exp` claim).

Note that if you are using the client created [from here](https://github.com/kkosmowski/next-social-app/blob/e642a0adc4145a50497765e1497a7d15fb472d3b/src/app/api/pocketbase.ts#L5) in a server-side context, then this is wrong because requests from different users will write over the same `pb.authStore` state (`pb` in the above is a global variable). When using JS SSR in a long running node process, you need a unique SDK instance for each request (_Next.js is especially difficult for this because the last time I checked there wasn't an easy way to share complex request app state/objects between middlewares_). For general SSR recommendations, please check [SSR integration](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#ssr-integration).


If you still think that this is a PocketBase issue, then please create a **minimal** reproducible example and I'll try to have a look at it. One of the easiest way to rule out PocketBase or JS SDK issue is to create a static `pb_public/index.html` file:

```html
<!DOCTYPE html />
<html>
<head>
    <meta charset="utf-8" />
    <title>Test</title>
</head>
<body>
    <script type="module">
        import PocketBase from "https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@0.21.1/dist/pocketbase.es.mjs";

        const pb = new PocketBase("http://127.0.0.1:8090");

        await pb.collection("users").authWithPassword("test@example.com", "1234567890")

        console.log(pb.authStore.isValid)
    </script>
</body>
</html>
```

Navigate to `http://127.0.0.1:8090` in your browser and check the console.

**Reply by kkosmowski** [reply]
Looks like that was the case. It took me some time to refactor the auth, but now it works great.

I did not yet test it when there are multiple users logged in at the same time, but I took your good point and tried to make sure this "session client" will not share it between users.

For future references and in general, I leave some details:

Each time I make a request or even return a server rendered page, I create new PocketBase instance, load data from cookies, check if `authStore.isValid`. 

I export the authState to cookies only once, during sign in. I clear it when `isValid` is no longer true or on logout. I add 2 cookies, httpOnly `pb_auth` and "public" `is_logged` that informs client side whether it should call for `/me`.

I keep the data in the session client just in case, for now it is not used outside the `refreshData` method. Also, I  return `pb` together with data, to ensure the same PocketBase instance in case I want to follow the data check (for example update user's post or comment).

I did not implement the token refresh flow yet, but that's story for another day.
So far it has never logged me out like it used to, since the refactor. In case someone is interested in the changes I made, here's link to the commit:
https://github.com/kkosmowski/next-social-app/commit/d0fb029e07ec043b4dcda98d7df16362dd964442

Cheers @ganigeorgiev and really cool job on this tool 👍

**Comment by celestialmk** [comment]
I happen to be getting a token but pb.authStore.isValid returns false, the above correct answer works great in a test html page. I am using nativescript-vue

**Reply by ganigeorgiev** [reply]
I'm not familiar with nativescript-vue and I'm not sure I can help you with it, but if you are not able to resolve your issue, please file a new Q&A discussion elaborating a little more on the problem and ideally provide a **minimal** reproducible repo or at least a code sample and I'll try to have a look at it.
